In [1]:
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits import mplot3d
import pandas as pd
import wlcave as wlcave #need to path correctly!

C:\Users\tmhab\anaconda3\lib\site-packages\numpy\core\getlimits.py:174: RuntimeWarning: divide by zero encountered in exp2
  eps=exp2(ld(-112)),
C:\Users\tmhab\anaconda3\lib\site-packages\numpy\core\getlimits.py:53: RuntimeWarning: divide by zero encountered in log10
  self.precision = int(-log10(self.eps))
C:\Users\tmhab\anaconda3\lib\site-packages\numpy\core\getlimits.py:54: RuntimeWarning: divide by zero encountered in power
  self.resolution = float_to_float(float_conv(10) ** (-self.precision))
C:\Users\tmhab\anaconda3\lib\site-packages\numpy\core\getlimits.py:186: RuntimeWarning: divide by zero encountered in exp2
  epsneg_f80 = exp2(ld(-64))
C:\Users\tmhab\anaconda3\lib\site-packages\numpy\core\getlimits.py:187: RuntimeWarning: divide by zero encountered in exp2
  tiny_f80 = exp2(ld(-16382))
C:\Users\tmhab\anaconda3\lib\site-packages\numpy\core\getlimits.py:201: RuntimeWarning: divide by zero encountered in exp2
  eps=exp2(ld(-63)),
C:\Users\tmhab\anaconda3\lib\site-packages\nump

## Import Files From Simulation:

In [2]:
file = "C:\\users/tmhab/wlcsim_membrane/data/2.10.21_NoInteractions_2hrs_r11v0.txt"
data = np.loadtxt(file) #D:\\myfiles\welcome.txt


## Our Approach to Calculating Percolation Paths Through a Polymer Membrane:

### Defining The System:

In our system we have random copolymers with solvent molecules. Polymers are represented by beads of types 1 and 2 and solvent molecules are individual beads of type 0. No bead can overlap another. 

### Defining Percolation:

A percolation is a pathway through the polymer network that solvent molecules can move through. To form a percolation a solvent molecule will need to have at least one other solvent molecule adjacent to it within some path radius $r$

### What are we trying to learn from the system:

The goal in developing this code is to determine quantitatively the amount of percolation in a membrane. 

### Models for Percolation:
**1. Fire Spread Model:** In this model we measure the percolation paths by tracing the spots where the solvent molecules continue to be adjacent to the previous... It's easier to explain with a picture

## What we need to calculate:

Starting with the final equilibrium state we have x,y,z postion of a bead and it's identity (and also methylation state). We want to calculate percolation paths which will be represented by root mean square. 

#### What Percolation Model do we want to use:

1. Fire spead model:

Proceedure:
1. Pick a bead type
2. pick a random bead of that bead type
3. Check surround bead types to see if they are the same
    a. Pick a radius to search in


# Note: for some reason if you declare two different PolymerNetworks, they can interfere w/ each other (change values)

In [3]:
def generate_test_data(length=10,width=10,height=10,datapoints=500):
    X = np.random.random(datapoints)*length
    Y = np.random.random(datapoints)*width
    Z = np.random.random(datapoints)*height
    Beads = np.random.randint(3,size=(datapoints)).astype('uint8')
    Beads
    return X,Y,Z,Beads

#generate_test_data()

In [4]:
class Bead:
    '''
    '''
    def __init__(self, x, y, z, bead_type):
        self.x = x
        self.y = y
        self.z = z
        self.bead_type = bead_type
        


In [5]:
class Polymer:
    '''
    '''
    
    def __init__(self, bead_list):
        self.bead_list = bead_list
    
    def print_bead_list(self):
        for i, bead in enumerate(self.bead_list):
            print("Bead " + str(i) + ": " + str(bead.x) + ", " + str(bead.y) + ", " + str(bead.z) + " Type: " + str(bead.bead_type))
   

In [6]:
class PolymerNetwork:
    """
    This represents the postions of bead and their type (time invariant)
    
    Parameters
    ----------
    X : float [array] X 
    """
    
    def __init__(self,X,Y,Z, Beads, beads_per_polymer=40, persistence_length=2.03, bead_list=[], polymer_list=[]):
        self.X = X
        self.Y = Y
        self.Z = Z
        self.Beads = Beads
        self.beads_per_polymer = beads_per_polymer
        self.bead_list = bead_list
        self.polymer_list = polymer_list
        self.persistence_length = persistence_length
        
        self.density = Beads.shape[0]/((max(X)-min(X)) * (max(Y)-min(Y)) * (max(Z)-min(Z)))
        
    def visualize(self):
        fig = plt.figure()
        ax = plt.axes(projection='3d')
        ax.scatter3D(self.X[Beads==0], self.Y[Beads==0], self.Z[Beads==0],color='y')
        ax.scatter3D(self.X[Beads==1], self.Y[Beads==1], self.Z[Beads==1],color='r')
        ax.scatter3D(self.X[Beads==2], self.Y[Beads==2], self.Z[Beads==2],color='b')
        
    def nearest_neighbor(self,bead_num,radius):
        '''
        Find the average number of beads that surround a bead of the same type within the radius specified
        '''
        X_bead = self.X[Beads==bead_num]
        Y_bead = self.Y[Beads==bead_num]
        Z_bead = self.Z[Beads==bead_num]
        
        total_neighbors = 0
        for i in range(Beads[Beads==bead_num].shape[0]):
            point = np.array([X_bead[i],Y_bead[i],Z_bead[i]])
            displacement = np.sqrt((X_bead - point[0])**2 + (Y_bead - point[1])**2 + (Z_bead - point[2])**2)
            num_neighbor_points = displacement[displacement<=radius].shape[0] -1
            
            total_neighbors += num_neighbor_points

        return total_neighbors/(i+1)
    
    def create_beads(self):
        if len(self.bead_list) != 0:
            raise Exception("bead list already populated")
        for i in range(len(self.X)):
            bead = Bead(self.X[i], self.Y[i], self.Z[i], self.Beads[i])
            self.bead_list.append(bead)
    
    def print_bead_list(self):
        for i, bead in enumerate(self.bead_list):
            print("Bead " + str(i) + ": " + str(bead.x) + ", " + str(bead.y) + ", " + str(bead.z) + " Type: " + str(bead.bead_type))
    
    def create_polymers(self):
        if len(self.polymer_list) != 0:
            raise Exception("polymer list already populated")
        if len(self.bead_list) == 0:
            raise Exception("Need to run create_beads before create_polymers")
        for i in range(self.beads_per_polymer, len(self.bead_list) + 1, self.beads_per_polymer):
            polymer = Polymer(self.bead_list[i - self.beads_per_polymer : i])
            self.polymer_list.append(polymer)
        
    def print_polymer_list(self, verbose=False):
        for i, poly in enumerate(self.polymer_list):
            print("POLYMER " + str(i))
            if verbose:
                poly.print_bead_list()

In [7]:
'''
Determine the mean square interbead distance
Input: PolymerNetwork
       int skip_parameter: ratio of how many non-overlapping jumps of size d to skip when calculating the mean square 
       displacemnt between two beads distance d away. 1 means include all non-overlapping jumps,
       2 means every other, 3 means every third, etc.
Return: A DataFrame where the columns are a specific polymer in the network, rows are the number of "jumps" n,
        and the values are the average distance squared bewteen beads n jumps away (averaged over all different jumps 
        of size n possible in the polymer)
     <R^2> array of mean square end-to-end distances for all polymers with 0 to num_beads - 1 bonds
     non-dimensionalized by (2*l_p)^2
'''
def r2(polymer_network, skip_parameter=1):
    if len(polymer_network.bead_list) == 0:
        raise Exception("Need to run create_beads")
    if len(polymer_network.polymer_list) == 0:
        raise Exception("Need to run create_polymers")
    num_poly = len(polymer_network.polymer_list)
    avg_n_jumps = pd.DataFrame(index=range(polymer_network.beads_per_polymer), columns=range(num_poly))
    
    for i, poly in enumerate(polymer_network.polymer_list):
        # creating matrix of (inter-bead distances)^2 for a given polymer
        bead_dist = create_polymer_r2_table(poly)
        
        # get diagonals (non-overlapping jumps of size 0 - all beads), avg, store in the dataframe
        avg_n_jumps[i] = get_avg_jump_distances(poly, bead_dist, skip_parameter)
        
        if (i % 500 == 0):
            print("processed polymer " + str(i))
    
    #average the distance b/w chain ends, normailize (???)
    num_beads = polymer_network.beads_per_polymer
    r2_ave_vect = np.zeros(num_beads)
    for i in range(num_beads):
        row_avg = np.average(avg_n_jumps.iloc[i]) # average size i jump over all polymers
        r2_ave_vect[i] = row_avg
    
    #print("r2_ave before normalization = " + str(np.average(avg_n_jumps.iloc[num_beads-1])))
    #r2_ave = np.average(avg_n_jumps.iloc[num_beads-1])/(2 * polymer_network.persistence_length)**2
    
    return r2_ave_vect, avg_n_jumps

'''
Creates a 2D array with the square distance from each bead to each other bead in a polymer.
Inputs: Polymer class
Returns: 2D array of square distances
'''
def create_polymer_r2_table(polymer):
    num_beads = len(polymer.bead_list)
    bead_dist = np.zeros((num_beads, num_beads))    
    for j, bead1 in enumerate(polymer.bead_list):
        for k, bead2 in enumerate(polymer.bead_list):
            position1 = np.array([bead1.x, bead1.y, bead1.z])
            position2 = np.array([bead2.x, bead2.y, bead2.z])
            displacement_squared = np.sum((position2 - position1)**2)
            bead_dist[j][k] = displacement_squared
    return bead_dist

'''
Given a 2D array of square bead displacements, this function finds the average square distance of all 
non overlapping jumps, for all possible jump sizes. 
Inputs: Polymer
        polymer_bead_dist- 2D array of bead square displacements
        skip_parameter- ratio of how many non-overlapping jumps of size d to skip. 1 means include all non-overlapping
        jumps, 2 means every other, 3 means every third, etc.
Returns: n_jump_dist_avg- a 1D array where the index i corresponds to the average square displacement of a jump size i
'''
def get_avg_jump_distances(polymer, polymer_bead_dist, skip_parameter):
    num_beads = len(polymer.bead_list)
    n_jump_dist_avg = np.zeros(num_beads)
    for d in range(num_beads):   # for all jump sizes
            diag = np.diag(polymer_bead_dist, d) # get jumps of size d
            if d == 0:
                n_jump_dist_avg[d] = np.average(diag)
            else:
                n_jump_dist_avg[d] = np.average(diag[::skip_parameter*d])  # average all non-overlapping jumps
    return n_jump_dist_avg    

In [8]:
#TEST DATA:

#X,Y,Z,Beads = generate_test_data()
#PNet_Test = PolymerNetwork(X,Y,Z,Beads, beads_per_polymer = 10)
#PNet_Test.create_beads()
#PNet_Test.create_polymers()
#PNet_Test.density

In [9]:
#PNet_Test.visualize()

In [10]:
#PNet_Test.nearest_neighbor(bead_num=0,radius=0.4)

In [11]:
#radii = np.linspace(0,15,100)
#for r in radii:
#    plt.plot(r,PNet_Test.nearest_neighbor(bead_num=0,radius=r),'y.')
#plt.xlabel('Radius')
#plt.ylabel('Average Number of Neighbor Points')

In [12]:
#x = r2(PNet_Test)
#x  # roughly random dist (no polymer chain)

In [13]:
#REAL DATA:

beads = np.zeros(len(data[:, 0]))   # no chemical identity data
no_interactions = PolymerNetwork(data[:, 0], data[:, 1], data[:, 2], beads, beads_per_polymer=40)

C:\Users\tmhab\anaconda3\lib\site-packages\ipykernel_launcher.py:20: RuntimeWarning: divide by zero encountered in double_scalars


In [14]:
no_interactions.create_beads()
#no_interactions.print_bead_list()
len(no_interactions.bead_list)  # should be 80,000

80000

In [15]:
no_interactions.create_polymers()
#no_interactions.print_polymer_list(verbose=True)
print(len(no_interactions.polymer_list)) #should be 2000


2000


In [16]:
#no_interactions.visualize()  #too many beads?

In [17]:
r2_ave_vect, result = r2(no_interactions)

C:\Users\tmhab\anaconda3\lib\site-packages\pandas\core\dtypes\missing.py:262: RuntimeWarning: divide by zero encountered in isnan
  result = np.isnan(values)
C:\Users\tmhab\anaconda3\lib\site-packages\pandas\core\dtypes\missing.py:379: RuntimeWarning: divide by zero encountered in invert
  return ~res
C:\Users\tmhab\anaconda3\lib\site-packages\pandas\io\formats\format.py:1252: RuntimeWarning: divide by zero encountered in bitwise_and
  is_float_type = lib.map_infer(vals, is_float) & notna(vals)
C:\Users\tmhab\anaconda3\lib\site-packages\numpy\core\_methods.py:45: RuntimeWarning: divide by zero encountered in reduce
  return umr_any(a, axis, dtype, out, keepdims)
C:\Users\tmhab\anaconda3\lib\site-packages\pandas\core\internals\managers.py:226: RuntimeWarning: divide by zero encountered in equal
  if (new_blknos == -1).any():
C:\Users\tmhab\anaconda3\lib\site-packages\ipykernel_launcher.py:55: RuntimeWarning: divide by zero encountered in subtract
C:\Users\tmhab\anaconda3\lib\site-packag

processed polymer 0
processed polymer 500
processed polymer 1000
processed polymer 1500


C:\Users\tmhab\anaconda3\lib\site-packages\pandas\core\internals\managers.py:1110: RuntimeWarning: divide by zero encountered in invert
  new_blknos[~is_deleted] = np.arange(self.nblocks - len(removed_blknos))


In [18]:
print(r2_ave_vect)
result 

[ 0.          0.16151278  0.59243728  1.24522628  2.07954993  3.06723617
  4.16569789  5.38538837  6.6487218   7.99208484  9.43793932 10.91659667
 12.37818563 13.89308667 15.4362178  17.04347128 18.64487326 20.22891804
 21.86025942 23.50434944 24.94745151 26.62523187 28.29599453 29.94128328
 31.55428672 33.20252514 34.84755739 36.54806549 38.21162255 39.91101385
 41.60153843 43.23246742 44.83438119 46.43516301 48.01264577 49.62889591
 51.19779563 52.80529516 54.39301144 56.05036629]


C:\Users\tmhab\anaconda3\lib\site-packages\numpy\core\arrayprint.py:885: RuntimeWarning: divide by zero encountered in isfinite
  finite_vals = data[isfinite(data)]
C:\Users\tmhab\anaconda3\lib\site-packages\numpy\core\arrayprint.py:888: RuntimeWarning: divide by zero encountered in not_equal
  abs_non_zero = absolute(finite_vals[finite_vals != 0])
C:\Users\tmhab\anaconda3\lib\site-packages\numpy\core\arrayprint.py:888: RuntimeWarning: divide by zero encountered in absolute
  abs_non_zero = absolute(finite_vals[finite_vals != 0])
C:\Users\tmhab\anaconda3\lib\site-packages\numpy\core\arrayprint.py:894: RuntimeWarning: divide by zero encountered in double_scalars
  (min_val < 0.0001 or max_val/min_val > 1000.)):
C:\Users\tmhab\anaconda3\lib\site-packages\numpy\core\arrayprint.py:968: RuntimeWarning: divide by zero encountered in isfinite
  if not np.isfinite(x):
C:\Users\tmhab\anaconda3\lib\site-packages\pandas\core\internals\managers.py:226: RuntimeWarning: divide by zero encountered in

0          1          2          3          4          5     \
0    0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   
1    0.157959   0.147473   0.162123   0.171162   0.162561   0.162067   
2    0.574533   0.549160   0.584291   0.626571   0.616642   0.601353   
3    1.155976   1.179395   1.214368   1.318761   1.276532   1.260369   
4    2.002388   1.891087   2.092197   2.065103   2.271605   2.145827   
5    2.718192   2.596062   2.963646   3.143028   3.352720   3.179280   
6    3.841749   3.502546   4.018099   4.069769   4.222490   4.509687   
7    5.468836   3.956770   5.277126   4.600353   5.480911   5.530452   
8    6.759105   5.748485   7.181959   5.340157   6.071908   6.127135   
9    7.181217   6.738717   6.911374   8.573450   8.030048   8.277606   
10   8.278792   8.523152   9.652447   7.624057   9.147504   9.113007   
11   8.729630   9.918771  12.116382   6.616148  10.504722   9.811240   
12   8.169497  11.643576  10.294108   8.083144  12.132253  11.940640   
13   8.822991  12.548467   8.507233  12.076776  15.636253  12.328823   
14   8.047095  14.971541  10.402255   9.050711  13.431123  15.960330   
15   9.584435  15.719070  12.437033   7.078428  11.818775  18.963092   
16  12.684555  15.329071  14.367604   7.155255  13.307144  21.551928   
17  15.370694  16.773631  16.641538   6.756725  15.392058  23.790974   
18  15.145070  19.063472  16.097036   6.591769  17.662168  25.296666   
19  16.191203  19.809939  14.540148   6.415922  20.937653  23.156670   
20  11.910080  15.842253  12.465369   5.838478  19.070267  10.672119   
21  14.363671  16.723715  12.927286   6.102057  21.171561   9.139413   
22  16.555755  17.839114  13.745141   6.247583  22.196909   7.521140   
23  18.525492  18.969714  15.047524   6.905576  23.087386   6.764580   
24  20.007471  21.756753  15.922152   7.705862  24.754476   5.852385   
25  21.253199  23.611330  17.350558   9.722910  29.108575   5.346937   
26  23.026640  23.699019  18.320046  11.857797  33.627929   4.795669   
27  26.816372  23.916427  20.735008  13.094030  36.934192   4.731780   
28  30.529796  23.254504  23.296314  14.876158  38.988466   5.800852   
29  33.710382  22.874768  26.782924  14.740304  39.838918   8.007285   
30  36.470295  21.737942  28.550015  12.907286  39.334334   9.414191   
31  41.477783  18.831188  31.119810  11.118637  41.856960  10.824705   
32  47.006366  16.558949  33.077104   8.708677  42.503172  12.904927   
33  52.579197  16.191983  37.774445   6.560078  44.072397  14.153104   
34  56.137471  16.980044  40.558875   5.098563  46.159543  16.264483   
35  56.312188  19.670262  41.549178   3.200347  45.957057  17.364090   
36  55.799413  23.091864  40.812389   2.688615  44.278353  18.600706   
37  56.050495  25.549288  40.397073   3.196354  44.351662  17.713401   
38  58.600914  28.917652  38.514136   3.845801  46.899238  16.783446   
39  60.508278  32.721549  38.540707   5.371965  47.350437  14.931495   

         6          7          8          9     ...       1990       1991  \
0    0.000000   0.000000   0.000000   0.000000  ...   0.000000   0.000000   
1    0.161183   0.154522   0.170941   0.160492  ...   0.151643   0.160275   
2    0.595330   0.565054   0.620521   0.593195  ...   0.551825   0.585822   
3    1.230608   1.188685   1.311302   1.233289  ...   1.217028   1.286766   
4    2.012835   1.760094   2.125436   2.012604  ...   1.989987   2.051770   
5    2.946224   2.635450   3.060645   2.772549  ...   3.017031   3.008510   
6    3.687525   3.610766   4.294624   3.938173  ...   4.040649   3.386335   
7    5.266811   4.162948   5.190584   4.970776  ...   5.364221   5.052126   
8    6.328008   5.391625   6.886879   6.510991  ...   7.253022   5.963385   
9    7.343234   7.277459   7.836084   7.823021  ...   9.385375   6.737436   
10   9.879473   8.622556   6.806391   9.381960  ...   9.719786   9.206338   
11  12.074126   8.683771   6.936987  10.070343  ...  11.577030  10.542381   
12  11.840041   7.682601   9.030592  10.100547  ...  13

In [19]:
#wlc: kuhn length b ~= 2 * persistence length
kuhn_l = 2 * no_interactions.persistence_length
print(kuhn_l)

4.06


In [20]:
wlcave.r2_ave(kuhn_l, 3)

C:\Users\tmhab\wlcsim_membrane\wlcave.py:26: RuntimeWarning: divide by zero encountered in exp
  - (1 - np.exp(-(dimensions - 1) * length_kuhn)) / (dimensions - 1) ** 2)
C:\Users\tmhab\wlcsim_membrane\wlcave.py:26: RuntimeWarning: divide by zero encountered in double_scalars
  - (1 - np.exp(-(dimensions - 1) * length_kuhn)) / (dimensions - 1) ** 2)


3.5601487643302203

Not = to 3.4  :|

Plot the number of beads/size of jumps (row index) vs avg displacement squared (average of row) ?